specs/dwritecore/DWriteCore.md

```diff
@@ -1235,3 +2025,906 @@ DWRITE_BEGIN_INTERFACE(IDWriteFactory8, "202D44F0-84AC-4D6F-B8D0-704B27675CAA")
+        /// Specifies an affine transform to be applied to child content.
+        /// This paint element has one child element, which is the transformed content.
+        /// </summary>
+        /// <remarks>
+        /// This corresponds to paint formats 12 through 31 in the OpenType COLR table.
+        /// </remarks>
+        typedef DWRITE_MATRIX PAINT_TRANSFORM;
+        PAINT_TRANSFORM transform;
+
+        /// <summary>
+        /// Valid for paint elements of type DWRITE_PAINT_TYPE_COMPOSITE.
+        /// Combines the two child paint elements using the specified compositing or blending mode.
+        /// This paint element has two child elements. The first child is the paint source. The
+        /// second child is the paint destination (or backdrop).
+        /// </summary>
+        /// <remarks>
+        /// This corresponds to a PaintComposite record in the OpenType COLR table.
+        /// </remarks>
+        struct PAINT_COMPOSITE
+        {
+            /// <summary>
+            /// Specifies the compositing or blending mode.
+            /// </summary>
+            DWRITE_COLOR_COMPOSITE_MODE mode;
+        } composite;
+    } paint;
+};
+```
+
+### IDWritePaintReader Interface
+
+```c++
+/// <summary>
+/// Interface used to read color glyph data for a specific font. A color glyph is
+/// represented as a visual tree of paint elements.
+/// </summary>
+DWRITE_BEGIN_INTERFACE(IDWritePaintReader, "8128E912-3B97-42A5-AB6C-24AAD3A86E54") : IUnknown
+{
+    /// <summary>
+    /// Sets the current glyph and positions the reader on the root paint element of the
+    /// selected glyph's visual tree.
+    /// </summary>
+    /// <param name="glyphIndex">Glyph index to get the color glyph representation for.</param>
+    /// <param name="paintElement">Receives information about the root paint element of the
+    /// glyph's visual tree.</param>
+    /// <param name="structSize">Size of the DWRITE_PAINT_ELEMENT structure, in bytes.</param>
+    /// <param name="clipBox">Receives a precomputed glyph box (in ems) for the specified glyph,
+    /// if one is specified by the font. Otherwise, the glyph box is set to an empty rectangle
+    /// (all zeros). If a non-empty clip box is specified, the client must clip the color
+    /// glyph's representation to the specified box.</param>
+    /// <param name="glyphAttributes">Receives optional paint attributes for the glyph.</param>
+    /// <returns>
+    /// Standard HRESULT error code.
+    /// </returns>
+    /// <remarks>
+    /// If the specified glyph index is not a color glyph, the method succeeds, but the paintType
+    /// member of the DWRITE_PAINT_ELEMENT structure is set to DWRITE_PAINT_TYPE_NONE. In this
+    /// case, the application should draw the input glyph as a non-color glyph.
+    /// </remarks>
+    STDMETHOD(SetCurrentGlyph)(
+        UINT32 glyphIndex,
+        _Out_writes_bytes_(structSize) DWRITE_PAINT_ELEMENT* paintElement,
+        UINT32 structSize,
```

Comment:
Why not have this overload define the type of paintElement as void* ? It seems kind of weird for the type to be DWRITE_PAINT_ELEMENT, however the call itself will end up writing an arbitrary number of bytes into the out parameter. This is a logical implication that the type should not be set.

Reply:
Yes, I anticipate discussing this with the API review panel.

I agree it's a little weird, but it's not unprecedented. There are quite a few classic Win32 APIs in which the client is expected to set the `cbSize` field of a structure before passing it to a function that be filled in. This is to enable a structure to be versioned while ensuring we never write past the end if the app was compiled against an older version.

In this case, we may want to define new paint types in the future (for COLR v2 and later). I propose to just add those to the existing union instead of having to define a DWRITE_PAINT_ELEMENT1 along with new interfaces and methods to go with it. That _may_ mean the size of the structure increases. However, it should not break ABI compatibility because the DWRITE_PAINT_FEATURE_LEVEL ensures existing clients only see paint types they know about. The size parameter is just an extra measure to ensure we never write past the end of the structure.